File: all-submissions/0023-merge-k-sorted-lists/solution.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        Comparator<ListNode> comparator;
        comparator = new Comparator<ListNode>() {
            @Override
            public int compare(ListNode o1, ListNode o2) {
                return o1.val - o2.val;
            }
        };
        PriorityQueue<ListNode> priorityQueue = new PriorityQueue<ListNode>(comparator);
        for(ListNode list: lists) {
            if(list != null)
                priorityQueue.add(list);
        }

        ListNode head = new ListNode(0);
        ListNode curr = head;
        while(!priorityQueue.isEmpty()){
            curr.next = priorityQueue.poll();
            curr = curr.next;
            ListNode next = curr.next;
            if(next!=null){
                priorityQueue.add(next);
            }
        }
        return head.next;

        

    }
}

File: all-submissions/0023-merge-k-sorted-lists/solution.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        if not lists: 
            return None
        def merge2Lists(listA, listB):
            dummyHead = ListNode()
            curr = dummyHead
            while listA and listB:
                valA = listA.val
                valB = listB.val
                if valA<=valB:
                    curr.next = listA
                    listA = listA.next
                elif valA > valB:
                    curr.next = listB
                    listB = listB.next
                curr = curr.next
            curr.next = listA if not listB else listB
            return dummyHead.next

        
        while(len(lists)) > 1:
            temp = []
            for i in range(0, len(lists), 2):
                l1 = lists[i]
                l2 = lists[i+1] if i + 1 < len(lists) else None
                temp.append(merge2Lists(l1, l2))  
            lists = temp
        return lists[0]
